### ISSUE TITLE
mkdir/rmdir/unlink.t fails with perl 5.10.1

### ISSUE THREAD

**Issue by eserte** [CpanelInc/Test-MockFile#21]
See subject. A sample fail report: http://www.cpantesters.org/cpan/report/c3960358-e403-11e8-98d5-bb14e5798fec
This seems to happen only with perl 5.10.1.

**Comment by toddr** [comment]
Thank you. Any chance I can get verbose output from one of these systems? I'd like to know if BSD is giving a different error number or just failing to handle the corner case.

**Comment by eserte** [comment]
It's not only BSD, I see the failure also with Debian/jessie or centos6. The detailed test log (this is from a jessie system):
```
# Seeded srand with seed '20181114' from local date.
not ok 1 - basic mkdir {
    ok 1 - REAL mkdir when dir is missing.
    ok 2 -  - $! is unset.
    ok 3 - REAL rmdir when dir is there
    not ok 4 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 5 - MOCK mkdir when dir is missing.
    ok 6 -  - $! is unset.
    ok 7 - Perms are 0755
    not ok 8 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 9 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 10 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 11 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 12 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 13 - 1
    ok 14 -  - $! is unset.
    not ok 15 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 16 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 17 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 18 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 19 - MOCK rmdir when dir is there
    ok 20 -  - $! is unset.
    not ok 21 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 22 - 1
    not ok 23 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 24 - 1
    1..24
}
# Failed test 'basic mkdir'
# at t/mkdir.t line 39.
ok 2 - undef dir name {
    ok 1 - REAL mkdir when undef is passed as the file name.
    ok 2 -  - $! is ENOENT.
    ok 3 -  - Returns 0
    ok 4 - MOCK mkdir when undef is passed as the file name.
    ok 5 -  - $! is ENOENT.
    ok 6 -  - Returns 0
    1..6
}
not ok 3 - REAL mkdir {
    ok 1 - put the real tempdir back
    not ok 2 - A real mkdir through the shim
    # Failed test 'A real mkdir through the shim'
    # at t/mkdir.t line 59.
    # +-----+----+-------+
    # | GOT | OP | CHECK |
    # +-----+----+-------+
    # | 0   | eq | 1     |
    # +-----+----+-------+
    not ok 3 -  - $! is unset.
    # Failed test ' - $! is unset.'
    # at t/mkdir.t line 60.
    # +-----+----+-------+
    # | GOT | OP | CHECK |
    # +-----+----+-------+
    # | 17  | eq | 0     |
    # +-----+----+-------+
    # $! = File exists
    ok 4 - A real mkdir through the shim when it exists already
    ok 5 -  - $! is EEXIST.
    1..5
}
# Failed test 'REAL mkdir'
# at t/mkdir.t line 67.
not ok 4 - mkdir when file exists {
    ok 1 - A real mkdir when the dir is already a file.
    ok 2 -  - $! is EEXIST.
    ok 3 - A mock mkdir when the dir is already a file.
    ok 4 -  - $! is EEXIST.
    not ok 5 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 6 - A mock mkdir when the path is a mocked file but not on disk becomes a directory mock.
    ok 7 - $mock is now a directory
    1..7
}
# Failed test 'mkdir when file exists'
# at t/mkdir.t line 88.
not ok 5 - mkdir when symlink exists {
    ok 1 - A real mkdir when the dir is already a symlink.
    ok 2 -  - $! is EEXIST.
    ok 3 - A mock mkdir when the dir is already a symlink.
    ok 4 -  - $! is EEXIST.
    not ok 5 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 6 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 7 - A mock mkdir when the path is a mocked symlink but not on disk turns the mock object into a dir.
    ok 8 - $mock is now a directory
    1..8
}
# Failed test 'mkdir when symlink exists'
# at t/mkdir.t line 110.
not ok 6 - mkdir with file perms {
    ok 1 - A real mkdir with 0770 perms.
    ok 2 -  - $! is unset.
    ok 3 - permissions are the real file's permissions
    not ok 4 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 5 - A mock mkdir with 0700 perms.
    ok 6 -  - $! is unset.
    ok 7 - Permissions are the mock permissions of 0700
    not ok 8 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 9 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 10 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    not ok 11 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 12 - Remove the fake dir
    not ok 13 - Unexpected warning: Use of uninitialized value in length at /tmpfs/.cpan-build-cpansand/2018111407/Test-MockFile-0.011-0/blib/lib/Test/MockFile.pm line 713.
    ok 14 - A mock mkdir with 0700 perms.
    ok 15 -  - $! is unset.
    ok 16 - Permissions get umask applied.
    1..16
}
# Failed test 'mkdir with file perms'
# at t/mkdir.t line 136.
1..6
```

**Comment by eserte** [comment]
Here's something which can be used as a `Dockerfile` to reproduce the problem (the system perl in CentOS6 uses to be 5.10.1, so the problem also happens here):
```
# docker build -t perl-test . && docker run perl-test
FROM centos:6
RUN yum install -y perl-CPAN
RUN yum install -y perl-Time-HiRes
RUN yum install -y gcc
CMD cpan -t TODDR/Test-MockFile-0.011.tar.gz
```

**Comment by toddr** [comment]
Ugh! so sorry it was in my travis output and I wasn't paying attention 

https://travis-ci.org/CpanelInc/Test-MockFile/jobs/452532189

**Comment by toddr** [comment]
This was fixed in d167a2c112c972db05780ac5ccebe6897b21d241
